### PR TITLE
Add configurable auto-application of Gradle Enterprise plugin

### PR DESCRIPTION
**Pull request by clayburn**
This change adds configuration parameters to allow plugin users to
automatically apply and configure the Gradle Enterprise plugin from
TeamCity configuration parameters. `GRADLE_ENTERPRISE_URL` will set the
`gradleEnterprise.server` property. `GRADLE_ENTERPRISE_PLUGIN_VERSION`
will apply the specified version of the Gradle Enterprise plugin if not
already applied.

The change adds the `teamCityBuildScanPlugin.gradle.enterprise.url` and
`teamCityBuildScanPlugin.gradle.enterprise.plugin.version`
system properties to the injected init script in order to pass the
configuration parameters from TeamCity into Gradle.

Support for the CCUD Gradle plugin has been added, too.

### CODE REVIEW

**Comment by clayburn** [comment]
Draft for now so we can start reviewing the approach. I'd like to test on a couple more TC versions before moving out of draft.

File: agent/src/main/resources/build-scan-init.gradle [hunk 1]
```diff
@@ -5,11 +5,43 @@ if (GradleVersion.current() < GradleVersion.version('4.1')) {
     return
 }
 
+initscript {
+    def geVersion = System.getProperty('teamCityBuildScanPlugin.gradle.enterprise.plugin.version')
+    def atLeastGradle5 = GradleVersion.current() >= GradleVersion.version('5.0')
+    def ccudVersion = System.getProperty('teamCityBuildScanPlugin.ccud.version')
```
Comment (etiennestuder) [inline agent/src/main/resources/build-scan-init.gradle:11]: Move one line up.

File: README.md [hunk 2]
```diff
@@ -70,6 +70,18 @@ previously downloaded plugin `.zip` file.
 
 1. Find the links of the published build scans in the _Overview_ section of each TeamCity build.
 
+## Automatic Gradle Enterprise Plugin Setup
+
+1. In TeamCity, on the build configuration for which you want to apply Gradle Enterprise, create 3 configuration parameters:
```
[the remark this answers is not in the document]
Comment (etiennestuder) [reply README.md:75]: Let's keep going with the current approach for now, so we can ensure it all is fully done besides where the config comes from.

File: agent/src/main/resources/build-scan-init.gradle [hunk 3]
```diff
@@ -5,11 +5,43 @@ if (GradleVersion.current() < GradleVersion.version('4.1')) {
...
+    if (geVersion && !gradle.getParent()) {
+        repositories {
+            maven { url 'https://plugins.gradle.org/m2' }
+        }
+        dependencies {
+            classpath atLeastGradle5 ?
+                    "com.gradle:gradle-enterprise-gradle-plugin:$geVersion" :
+                    "com.gradle:build-scan-plugin:$geVersion"
+
+            if (atLeastGradle5 && ccudVersion) {
```
Comment (etiennestuder) [inline agent/src/main/resources/build-scan-init.gradle:22]: I would do this check not nested. This way, you could still apply the CCUD plugin even if the GE plugin is not defined as a config param (for scenarios where the GE plugin is already defined in the build).

File: agent/src/main/resources/build-scan-init.gradle [hunk 4]
```diff
@@ -5,11 +5,43 @@ if (GradleVersion.current() < GradleVersion.version('4.1')) {
...
+
+def CCUD_PLUGIN_ID = 'com.gradle.common-custom-user-data-gradle-plugin'
+def CCUD_PLUGIN_CLASS = 'com.gradle.CommonCustomUserDataGradlePlugin'
+
+def GRADLE_ENTERPRISE_PLUGIN_ID = 'com.gradle.enterprise'
+def GRADLE_ENTERPRISE_PLUGIN_CLASS = 'com.gradle.enterprise.gradleplugin.GradleEnterprisePlugin'
 def GRADLE_ENTERPRISE_EXTENSION_CLASS = 'com.gradle.enterprise.gradleplugin.GradleEnterpriseExtension'
 
+def geUrl = System.getProperty('teamCityBuildScanPlugin.gradle.enterprise.url')
+def autoApplyGe = System.getProperty('teamCityBuildScanPlugin.gradle.enterprise.plugin.version') && !gradle.getParent()
```
Comment (etiennestuder) [inline agent/src/main/resources/build-scan-init.gradle:42]: Rename to `autoApplyGePlugin`.

File: agent/src/main/resources/build-scan-init.gradle [hunk 5]
```diff
@@ -5,11 +5,43 @@ if (GradleVersion.current() < GradleVersion.version('4.1')) {
...
+def autoApplyCcud = System.getProperty('teamCityBuildScanPlugin.ccud.version') && GradleVersion.current() >= GradleVersion.version('5.0')
```
Comment (etiennestuder) [inline agent/src/main/resources/build-scan-init.gradle:43]: Rename to `autoApplyCcudPlugin`.

File: agent/src/main/resources/build-scan-init.gradle [hunk 6]
```diff
@@ -21,21 +53,49 @@ def buildScanPublishedAction = { def buildScan ->
...
         }
     }
 }
 
 static def extensionsWithPublicType(def container, String publicType) {
-    container.extensions.extensionsSchema.elements.findAll { it.publicType.concreteClass.name == publicType }
+    container.extensions.extensionsSchema.elements
+            .findAll { it.publicType.concreteClass.name == publicType }
+            .collect { container[it.name] }
 }
```
Comment (etiennestuder) [inline agent/src/main/resources/build-scan-init.gradle:99]: Can you explain why this change was needed?

File: agent/src/main/resources/build-scan-init.gradle [hunk 7]
```diff
@@ -21,21 +53,49 @@ def buildScanPublishedAction = { def buildScan ->
...
+        }
+
+        if (autoApplyCcud && !pluginManager.hasPlugin(CCUD_PLUGIN_ID)) {
+            pluginManager.apply(initscript.classLoader.loadClass(CCUD_PLUGIN_CLASS))
+        }
+
         pluginManager.withPlugin(BUILD_SCAN_PLUGIN_ID) {
             buildScanPublishedAction(buildScan)
+            buildScan {
+                afterEvaluate {
```
Comment (etiennestuder) [inline agent/src/main/resources/build-scan-init.gradle:67]: Having afterEvaluate nested inside buildScan seems a bit odd.

Comment (etiennestuder) [reply agent/src/main/resources/build-scan-init.gradle:67]: I think I moved move this entire afterEvaluate block to within the if statement above (autoApplyCcud && !pluginManager.hasPlugin(CCUD_PLUGIN_ID) - and wrap it again with `pluginManager.withPlugin(BUILD_SCAN_PLUGIN_ID)`

File: agent/src/main/resources/build-scan-init.gradle [hunk 8]
```diff
@@ -21,21 +53,49 @@ def buildScanPublishedAction = { def buildScan ->
...
+        if (autoApplyCcud && !settings.pluginManager.hasPlugin(CCUD_PLUGIN_ID)) {
+            settings.pluginManager.apply(initscript.classLoader.loadClass(CCUD_PLUGIN_CLASS))
+        }
+
+        extensionsWithPublicType(settings, GRADLE_ENTERPRISE_EXTENSION_CLASS).each { ext ->
+            buildScanPublishedAction(ext.buildScan)
+            ext.server = geUrl ?: ext.server
+            ext.buildScan {
+                publishAlwaysIf(autoApplyGe)
+            }
```
Comment (etiennestuder) [inline agent/src/main/resources/build-scan-init.gradle:90]: I would do as I suggested above re: afterEvaluate and use the settingsEvaluated callback.

**Review by etiennestuder — CHANGES_REQUESTED** [review]

**Comment by etiennestuder** [comment]
@clayburn I've added some feedback. Mostly renamings. On the init script, I suggest we apply the following pattern:
- allow to apply the ge plugin and the ccud plugin independently of each other (i.e. no nested if between the two)
- do not make the application of the ge plugin or ccud plugin dependent on whether the ge url config was also provided
- only set the ge url if the ge plugin is actually applied

The above should be in line with the comments I left on the PR.

Let's tackle the global config separately.

File: agent/src/main/resources/build-scan-init.gradle (hunk 7 above)
Comment (clayburn) [reply agent/src/main/resources/build-scan-init.gradle]: Did you mean in the `if(autoApplyGePlugin...` block? That's what I did, since it seemed to make most sense to configure GE is we applied the GE plugin, not CCUD.

File: agent/src/main/resources/build-scan-init.gradle (hunk 6 above)
Comment (clayburn) [reply agent/src/main/resources/build-scan-init.gradle]: It wasn't, and I reverted on this update. I think this stuck around through some earlier iterations.

**Comment by etiennestuder** [comment]
Thanks, @clayburn !
